### PR TITLE
Add SQL load script to the hot deployment watched files list

### LINKED ISSUE #3661: Hibernate SQL load script is not reloaded during dev mode
**Describe the bug**
When updating the Hibernate SQL load script (e.g. default `import.sql`) during dev mode the changes are not reloaded after a new request

**Expected behavior**
Using drop-and-create I would expect the changes to the Hibernate SQL load script to trigger a reload and changes relected 

**Actual behavior**
Nothing happens and the creating or updating the import.sql doesn't trigger a reload of the application

**To Reproduce**
1. Run the following in a terminal to create a project using Panache Entity and RESTEasy
```sh
mvn io.quarkus:quarkus-maven-plugin:0.21.1:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=bug-report \
    -DclassName="org.acme.HelloResource" \
    -Dpath="/hello" \
    -Dextensions="quarkus-hibernate-orm-panache, resteasy-jsonb,quarkus-resteasy-jsonb,quarkus-jdbc-h2"

cd bug-report

cat <<EOF > src/main/resources/application.properties
quarkus.datasource.url=jdbc:h2:mem:default
quarkus.datasource.driver=org.h2.Driver
quarkus.datasource.username=sa
quarkus.hibernate-orm.database.generation=drop-and-create
quarkus.hibernate-orm.log.sql=true
EOF

cat <<EOF > src/main/java/org/acme/MyEntity.java
package com.acme;

import javax.persistence.Entity;

import io.quarkus.hibernate.orm.panache.PanacheEntity;

@Entity
public class MyEntity extends PanacheEntity {
  public String name;
  public boolean isThisWorking;
}
EOF

cat <<EOF > src/main/java/org/acme/MyEntity.java
package org.acme;

import javax.persistence.Entity;

import io.quarkus.hibernate.orm.panache.PanacheEntity;

@Entity
public class MyEntity extends PanacheEntity {
  public String name;
  public boolean isThisWorking;
}
EOF

cat <<EOF > src/main/java/org/acme/MyResource.java
package org.acme;

import java.util.List;

import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/api")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class MyResource {

  @GET
  public List<MyEntity> listAll() {
    return MyEntity.listAll();
  }
}
EOF
```
2. Start Quarkus in dev mode `mvn compile quarkus:dev`
3. In a separate terminal test the end point using HTTPie or curl 
```sh
http :8080/api
```
4. Now create a import.sql with the following command
```sh
cat <<EOF > src/main/resources/import.sql
INSERT INTO MyEntity(id,name,isThisWorking) VALUES (nextval('hibernate_sequence'), 'Tester', false);
EOF
```

5. Test the endpoint again
```sh
http :8080/api
```

Here we should see a JSON array with the entry in it, but the array is empty. Restarting quarkus solves the issue.

**Configuration**
```properties
quarkus.datasource.url=jdbc:h2:mem:default
quarkus.datasource.driver=org.h2.Driver
quarkus.datasource.username=sa
quarkus.hibernate-orm.database.generation=drop-and-create
quarkus.hibernate-orm.log.sql=true
```

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:
```
uname -a
Darwin Q2.lan 18.7.0 Darwin Kernel Version 18.7.0: Thu Jun 20 18:42:21 PDT 2019; root:xnu-4903.270.47~4/RELEASE_X86_64 x86_64
``` 
 - Output of `java -version`: 
```
java -version
openjdk version "11.0.2" 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
```
 - GraalVM version (if different from Java): 
N/A
 - Quarkus version or git rev: 
0.21.1

**Additional context**

### PR DESCRIPTION
**Pull request by gwenneg**
Fixes #3661

### CODE REVIEW

**Comment by gwenneg** [comment]
This fixes the bug in the issue reproducer, but I had some weird results locally with the devmode test. I want to see what CI has to say about it.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 1]
```diff
@@ -104,8 +105,9 @@
     HibernateOrmConfig hibernateConfig;
 
     @BuildStep
-    HotDeploymentWatchedFileBuildItem configFile() {
-        return new HotDeploymentWatchedFileBuildItem("META-INF/persistence.xml");
+    List<HotDeploymentWatchedFileBuildItem> hotDeploymentFiles() {
+        return Arrays.asList(new HotDeploymentWatchedFileBuildItem("META-INF/persistence.xml"),
+                new HotDeploymentWatchedFileBuildItem("import.sql"));
```
Comment (machi1990) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: `import.sql` should be replaced by the value read from the Hibernate Configuration. This piece of code was previously here - https://github.com/quarkusio/quarkus/blob/beac5f75ab0b67979e0de65b5107dd120ab0b3fb/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L440-L450

**Review by machi1990 — CHANGES_REQUESTED** [review]
Good catch @gwenneg, I just have one suggestion.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java (hunk 1 above)
Comment (gwenneg) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Good point, thanks @machi1990!

**Comment by gwenneg** [comment]
It seems that CI is happier with this than my local environment :)

I'll push the improved version then!

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 2]
```diff
@@ -104,8 +105,9 @@
     HibernateOrmConfig hibernateConfig;
 
     @BuildStep
-    HotDeploymentWatchedFileBuildItem configFile() {
-        return new HotDeploymentWatchedFileBuildItem("META-INF/persistence.xml");
+    List<HotDeploymentWatchedFileBuildItem> hotDeploymentFiles() {
+        return Arrays.asList(new HotDeploymentWatchedFileBuildItem("META-INF/persistence.xml"),
+                new HotDeploymentWatchedFileBuildItem(getSqlLoadScript()));
```
Comment (machi1990) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: This means will we also produce a `HotDeploymentWatchedFileBuildItem` if `no-file` is provided as  sql load script?

Comment (gwenneg) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Yes and that seems pretty pointless, I'm updating the PR again, thanks!

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 3]
```diff
@@ -122,7 +130,6 @@ public void build(RecorderContext recorderContext, HibernateOrmRecorder recorder
             BuildProducer<FeatureBuildItem> feature,
             BuildProducer<PersistenceUnitDescriptorBuildItem> persistenceUnitDescriptorProducer,
             BuildProducer<SubstrateResourceBuildItem> resourceProducer,
-            BuildProducer<HotDeploymentWatchedFileBuildItem> hotDeploymentProducer,
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Yeah, I think we used to use it and it got broken.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 4]
```diff
@@ -311,6 +318,15 @@ public void startPersistenceUnits(HibernateOrmRecorder recorder, BeanContainerBu
         recorder.startAllPersistenceUnits(beanContainer.getValue());
     }
 
+    private String getSqlLoadScript() {
+        // Explicit file or default Hibernate ORM file.
+        return hibernateConfig.sqlLoadScript.orElse("import.sql");
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: I would make this method return an `Optional` and take into account the `no-file` case directly. That would be cleaner.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 5]
```diff
@@ -223,7 +230,7 @@ void handleNativeImageImportSql(BuildProducer<SubstrateResourceBuildItem> resour
                 resources.produce(new SubstrateResourceBuildItem(
                         (String) i.getDescriptor().getProperties().get("javax.persistence.sql-load-script-source")));
             } else {
-                resources.produce(new SubstrateResourceBuildItem("import.sql"));
+                resources.produce(new SubstrateResourceBuildItem(getSqlLoadScript()));
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: And with using an `Optional` (see comment below) you could manage this properly too.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a couple of suggestions.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java (hunk 3 above)
Comment (machi1990) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: It great that we'll have a test for it now. Good job @gwenneg :-)

**Comment by gwenneg** [comment]
Is my last commit what you meant @gsmet? I'll stash once I have the confirmation.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 6]
```diff
@@ -311,6 +318,16 @@ public void startPersistenceUnits(HibernateOrmRecorder recorder, BeanContainerBu
...
     }
 
+    private Optional<String> getSqlLoadScript() {
+        // Explicit file or default Hibernate ORM file.
+        String script = hibernateConfig.sqlLoadScript.orElse("import.sql");
+        if (NO_SQL_LOAD_SCRIPT_FILE.equalsIgnoreCase(script)) {
+            return Optional.empty();
+        } else {
+            return Optional.of(script);
+        }
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: I would code it this way:
```
private Optional<String> getSqlLoadScript() {
    if (NO_SQL_LOAD_SCRIPT_FILE.equalsIgnoreCase(script)) {
        return Optional.empty();
    }

    return Optional.of(hibernateConfig.sqlLoadScript.orElse("import.sql"));
}
```

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 7]
```diff
@@ -419,32 +435,29 @@ private void handleHibernateORMWithNoPersistenceXml(
...
-                                desc.getProperties().setProperty(AvailableSettings.HBM2DDL_IMPORT_FILES, importFile);
+                                desc.getProperties().setProperty(AvailableSettings.HBM2DDL_IMPORT_FILES, importFile.get());
                                 desc.getProperties().setProperty(AvailableSettings.HBM2DDL_IMPORT_FILES_SQL_EXTRACTOR,
                                         MultipleLinesSqlCommandExtractor.class.getName());
 
                             });
 
                     //raise exception if explicit file is not present (i.e. not the default)
                     hibernateConfig.sqlLoadScript
-                            .filter(o -> !isNoFile)
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: So it's not related to this patch but I really think all this logic could be made simpler without all these lambdas everywhere.

Would you be interested in having a look? If not I will give it a try next week.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java [hunk 8]
```diff
@@ -311,6 +318,16 @@ public void startPersistenceUnits(HibernateOrmRecorder recorder, BeanContainerBu
         recorder.startAllPersistenceUnits(beanContainer.getValue());
     }
 
+    private Optional<String> getSqlLoadScript() {
+        // Explicit file or default Hibernate ORM file.
+        String script = hibernateConfig.sqlLoadScript.orElse("import.sql");
+        if (NO_SQL_LOAD_SCRIPT_FILE.equalsIgnoreCase(script)) {
+            return Optional.empty();
+        } else {
+            return Optional.of(script);
```
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Yes looks good to me.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java (hunk 7 above)
Comment (gwenneg) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: Sure, I'll try to simplify it. Should I include it in this PR or create another one?

Comment (gsmet) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: I would make it an additional commit of this very PR.

Comment (gsmet) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: My intuition is that good old if logic would be more readable but I might be wrong.

Comment (gwenneg) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java]: You weren't wrong at all :)

**Comment by gwenneg** [comment]
@gsmet This PR is ready for another review.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks!

**Pull request merged**